paimon-core/src/main/java/org/apache/paimon/operation/TagDeletion.java

```diff
@@ -0,0 +1,172 @@
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import static org.apache.paimon.operation.DeletionUtils.containsDataFile;
+import static org.apache.paimon.operation.DeletionUtils.indexDataFiles;
+import static org.apache.paimon.operation.DeletionUtils.readEntries;
+import static org.apache.paimon.operation.DeletionUtils.tryDeleteDirectories;
+
+/**
+ * Delete tag files. This class doesn't check changelog files because they are handled by {@link
+ * SnapshotDeletion}.
+ */
+public class TagDeletion {
+
+    private final FileIO fileIO;
+    private final FileStorePathFactory pathFactory;
+    private final ManifestList manifestList;
+    private final ManifestFile manifestFile;
+    private final IndexFileHandler indexFileHandler;
+
+    private final SnapshotManager snapshotManager;
+    private final List<Snapshot> taggedSnapshots;
+
+    private final Map<BinaryRow, Map<Integer, Set<String>>> skippedDataFiles;
+    private final Set<String> skippedManifests;
+
+    public TagDeletion(
+            FileIO fileIO,
+            Path tablePath,
+            FileStorePathFactory pathFactory,
+            ManifestList manifestList,
+            ManifestFile manifestFile,
+            IndexFileHandler indexFileHandler) {
+        this.fileIO = fileIO;
+        this.pathFactory = pathFactory;
+        this.manifestList = manifestList;
+        this.manifestFile = manifestFile;
+        this.indexFileHandler = indexFileHandler;
+
+        this.snapshotManager = new SnapshotManager(fileIO, tablePath);
+        this.taggedSnapshots = new TagManager(fileIO, tablePath).taggedSnapshots();
+
+        this.skippedDataFiles = new HashMap<>();
+        this.skippedManifests = new HashSet<>();
+    }
+
+    /** Delete unused data files, manifest files and empty data file directories of tag. */
+    public void delete(Snapshot taggedSnapshot) {
+        // collect from the earliest snapshot's base
+        Snapshot earliest = snapshotManager.snapshot(snapshotManager.earliestSnapshotId());
+        collectSkip(earliest.baseManifests(manifestList));
+
+        // collect from the neighbor tags
+        int index = findIndex(taggedSnapshot);
+        if (index - 1 >= 0) {
+            collectSkip(taggedSnapshots.get(index - 1).dataManifests(manifestList));
+        }
+        if (index + 1 < taggedSnapshots.size()) {
+            collectSkip(taggedSnapshots.get(index + 1).dataManifests(manifestList));
+        }
+
+        // delete data files and empty directories
+        deleteDataFiles(taggedSnapshot);
+        // delete manifests
+        deleteManifestFiles(taggedSnapshot);
+    }
+
+    private void collectSkip(List<ManifestFileMeta> manifestFileMetas) {
+        Iterable<ManifestEntry> entries = readEntries(manifestFileMetas, manifestFile);
+        indexDataFiles(skippedDataFiles, ManifestEntry.mergeEntries(entries));
```

Comment:
`readEntries` and `indexDataFiles` always together.
You can introduce a method to do two things.